mission_control/src/mission.cpp

```diff
@@ -36,359 +36,44 @@
-        }
-      }
-    }
-  }
-  catch (...)
-  {
-    ROS_WARN("Mission Execution failed because of an execption");
-    AbortMission();
-  }
-
-  ROS_INFO("Leaving ProcessMission");
-}
-
-void Mission::AbortMission()
-{
-  if ((GetState() == MissionState::ABORTING) || (GetState() == MissionState::COMPLETE)) return;
+BT::NodeStatus Mission::getMissionStatus() { return missionStatus_; }
 
-  ROS_WARN("Aborting mission");
-  SetState(MissionState::ABORTING);
+std::string Mission::getCurrentMissionDescription() { return missionDescription_; }
 
-  m_threadAbortMission = boost::shared_ptr<boost::thread>(
-      new boost::thread(boost::bind(&Mission::processAbort, this)));
-}
+std::string Mission::getCurrentBehavioralName() { return behaviorName_; }
 
-void Mission::processAbort()
+NodeStatus Mission::executeMission()
 {
-  m_threadProcMission->interrupt();
-  m_threadProcMission->join();
-
-  current_abort_behavior_id = 0;
-  current_abort_behavior_name = "";
-
-  bool abortfirsttime = true;
-
-  try
-  {
-    while (1)
-    {
-      // Get the next abort behavior
-      boost::mutex::scoped_lock lock(m_mutCallbacks);
-      m_current_behavior = getNextAbortBehavior(abortfirsttime);
-      lock.unlock();
-      abortfirsttime = false;
-
-      if (m_current_behavior == NULL)
-      {
-        SetState(MissionState::COMPLETE);
-        ROS_INFO("Done with abort behaviors");
-        break;  // no more behaviors
-      }
-
-      current_abort_behavior_id++;
-      current_abort_behavior_name = m_current_behavior->getXmlTag();
-
-      ROS_INFO("Starting behavior [%s]", m_current_behavior->getXmlTag());
-      m_current_behavior->startBehavior();
-
-      // Sleep until we need to execute the behavior
-      int success = 0;
-      try
-      {
-        success = m_current_behavior->WaitForExecutionTimeSlot();
-      }
-      catch (boost::thread_interrupted)
-      {
-        break;
-      }
-
-      if (success == -1)
-      {
-        continue;  // go on to the next abort behavior
-      }
-
-      int behaviorTime = m_current_behavior->computeExecutionTimeForBehavior();
-      if (behaviorTime == -1)
-      {
-        continue;  // go on to the next abort behavior
-      }
-
-      m_current_behavior->ExecuteBehavior(node_handle);
-
-      unsigned int loop_delay_usec = 1000;
-      float time_elapsed = 0;
-      while (!m_current_behavior->getBehaviorDone())
-      {
-        usleep(loop_delay_usec);
-        int behaviorDuration = m_current_behavior->getBehaviorDuration();
-        if (behaviorDuration == 0.0) break;
-        if (behaviorDuration > 0.0)
-        {
-          time_elapsed += loop_delay_usec / 1000000.0;  // add seconds passed
-          if (time_elapsed >= behaviorDuration) break;
-        }
-      }
-    }
-  }
-  catch (...)
-  {
-    ROS_WARN("Mission Abort Execution failed because of an execption");
-  }
-
-  ROS_INFO("LEaving processAbort");
+  missionStatus_ = missionTree_.tickRoot();
+  return missionStatus_;
```

Comment:
@patapufate collapse in one line? Also, why not querying the tree's root node status instead of duplicating state?

Reply:
Fixed! Removing missionStatus variable and use rootNode()->status() instead